### PR TITLE
[Serialization] Refactor how we decide to load transitive dependencies

### PR DESCRIPTION
**Pull request by xymus**
Refactor and centralize the logic about when implementation-only and package-only dependencies should be loaded. Each transitive dependency is either required (like a normal import), ignored (like a transitive implementation-only dependency) or optional usually for the debugger that benefits from loading transitive implementation-only dependencies as it can use the information.

The new service `getTransitiveLoadingBehavior` is use from the module loading logic, the related diagnostics and from the dependency scanner.

This is mostly NFC, except that it fixes a few issues around package-only imports.

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Review by artemcm — APPROVED** [review]
Very nice to have it all in one place!

**Comment by xymus** [comment]
@swift-ci Please smoke test Windows

File: lib/Serialization/ModuleFile.cpp [hunk 1]
```diff
@@ -230,11 +218,13 @@ Status ModuleFile::associateWithFileContext(FileUnit *file, SourceLoc diagLoc,
           modulePath.front().Item == file->getParentModule()->getName()) {
         return error(Status::MissingUnderlyingModule);
       }
-
       // Otherwise, continue trying to load dependencies, so that we can list
       // everything that's missing.
```
Comment (bnbarham) [inline lib/Serialization/ModuleFile.cpp:222]: I think this makes more sense after the new `if` now?

**Review by bnbarham — APPROVED** [review]

File: lib/Serialization/ModuleFile.cpp (hunk 1 above)
Comment (xymus) [reply lib/Serialization/ModuleFile.cpp:222]: That comment is about the previous block and continues the comment before it `// If we're missing the module we're an overlay for, treat that specially.`. I moved the empty line around to make that more obvious, I think it helps a bit in the final result.

**Pull request merged**